c/ckb_mmr.h

```diff
@@ -0,0 +1,474 @@
+  _mmr_stack_value_t stack[MMR_STACK_SIZE];
+
+  uint32_t stack_top = 0;
+  _mmr_peak_t next_peak_info = _mmr_left_peak_height_pos(mmr_size);
+  uint64_t last_leaf_pos = 0;
+  int has_last_leaf = 0;
+  uint8_t command = 0xFF;
+
+  // We won't bother doing anything when mmr_size is 0.
+  if (mmr_size == 0) {
+    return ERROR_INVALID_PROOF;
+  }
+
+  while (1) {
+    int ret = MMR_COMMAND_READER(proof_reader_context, &command);
+    if (ret == ERROR_NO_MORE_COMMANDS) {
+      break;
+    }
+    if (ret != 0) {
+      return ret;
+    }
+    switch (command) {
+    case 1: {
+      if (stack_top >= MMR_STACK_SIZE) {
+        return ERROR_INVALID_STACK;
+      }
+      int leaf_success =
+          MMR_LEAF_READER(leaf_reader_context, &stack[stack_top].node,
+                          &stack[stack_top].position);
+      if (leaf_success != 0) {
+        return leaf_success;
+      }
+      if (has_last_leaf) {
+        if (last_leaf_pos >= stack[stack_top].position) {
+          return ERROR_INVALID_PROOF;
+        }
+      }
+      if (stack[stack_top].position >= mmr_size) {
+        return ERROR_INVALID_PROOF;
+      }
+      if (_mmr_pos_height_in_tree(stack[stack_top].position) > 0) {
+        return ERROR_INVALID_PROOF;
+      }
+      stack[stack_top].t = _MMR_STACK_NODE;
+      last_leaf_pos = stack[stack_top].position;
+      has_last_leaf = 1;
+      stack[stack_top].height = 0;
+      stack_top++;
+    } break;
+    case 2: {
+      if (stack_top >= MMR_STACK_SIZE) {
+        return ERROR_INVALID_STACK;
+      }
+      int ret = MMR_PROOF_READER(proof_reader_context, &stack[stack_top].node);
+      if (ret != 0) {
+        return ret;
+      }
+      stack[stack_top].t = _MMR_STACK_PROOF;
+      stack[stack_top].position = 0;
+      stack[stack_top].height = 0;
+      stack_top++;
+    } break;
+    case 3: {
+      if (stack_top < 2) {
+        return ERROR_INVALID_STACK;
+      }
+      uint32_t lhs_index = stack_top - 2;
+      uint32_t rhs_index = stack_top - 1;
+      uint64_t pos, next_pos;
+      uint8_t next_t;
+      uint32_t height;
+      const mmr_node_t *item;
+      const mmr_node_t *sibling_item;
+      if (stack[lhs_index].t == _MMR_STACK_PROOF) {
+        pos = stack[rhs_index].position;
+        height = stack[rhs_index].height;
+        next_pos = stack[lhs_index].position;
+        next_t = stack[lhs_index].t;
+        item = &stack[rhs_index].node;
+        sibling_item = &stack[lhs_index].node;
+      } else {
+        pos = stack[lhs_index].position;
+        height = stack[lhs_index].height;
+        next_pos = stack[rhs_index].position;
+        next_t = stack[rhs_index].t;
+        item = &stack[lhs_index].node;
+        sibling_item = &stack[rhs_index].node;
+      }
+      uint32_t next_height = _mmr_pos_height_in_tree(pos + 1);
+      uint64_t sibling_offset = _mmr_sibling_offset(height);
+      uint64_t sib_pos, parent_pos;
+      if (next_height > height) {
+        sib_pos = pos - sibling_offset;
```

Comment:
I think this is a similar case to the above one: `pos` is a user input indeed but we have tested to make sure 1) `pos` is a leaf 2) `pos` is less than `mmr_size`. `height` is actually a deduced field that starts from 0, and grow by 1 programmatically for each merge command, it is not a direct user input. So the question is: if the program logic is correct, so we have a valid leaf `pos` and it corresponding `height` starting from 0, can the `sibling_offset`, which is actually `(2 << height) - 1`, outgrow `pos`?

Reply:
When `pos` is 0, `next_height` will also be 0, since `height` is also 0 for a just-inserted leaf, the `if` statement on line 403 will be false, `pos - sibling_offset` will not be executed.